Lib/ufo2ft/featureWriters/markFeatureWriter.py

```diff
@@ -426,6 +457,62 @@ def _makeMarkToBaseAttachments(self):
             result.append(MarkToBasePos(glyphName, baseMarks))
         return result
 
+    def _makeGroupedMarkToBaseAttachments(self):
+        """Make markToBase attachements, then group them so that no group
+        contains conflicting anchor classes for the same glyph.
+        """
+        attachments = self._makeMarkToBaseAttachments()
+        # Idea: attachments is a list of mark to base pairs, linked together through an anchor name
+        # We have to put them into one or more lookups with the constraint
+        # that the same mark glyph cannot appear twice in the same lookup while
+        # using different anchor names.
+        markGlyphToMarkClasses = defaultdict(set)
+        for attachment in attachments:
+            for namedAnchor in attachment.marks:
+                for markGlyph in namedAnchor.markClass.glyphs:
+                    markGlyphToMarkClasses[markGlyph].add(namedAnchor.markClass.name)
+        # To compute the number of lookups that we need to build, we want
+        # the minimum number of lookups such that, whenever a mark glyph
+        # belongs to several mark classes, these classes are not in the same
+        # lookup. A trivial solution is to make 1 lookup per mark class
+        # but that's a bit wasteful, we might be able to do better by grouping
+        # mark classes that do not conflict.
+        # This is a graph coloring problem: the graph nodes are mark classes,
+        # edges are between classes that would conflict and the colors are
+        # the lookups in which they can go.
+        adjacency = {
+            # We'll get the same markClass several times in the dict
+            # comprehension below but it's ok, only one will be kept.
+            markClass: set()
+            for markClasses in markGlyphToMarkClasses.values()
+            for markClass in markClasses
+        }
+        for markGlyph, markClasses in markGlyphToMarkClasses.items():
+            for markClass in markClasses:
+                for other in markClasses:
+                    if other != markClass:
+                        adjacency[markClass].add(other)
+        colorGroups = colorGraph(adjacency)
+        # Sort the groups for reproducibility
```

Comment:
The order of the lookups should matter, I'm not sure it's a good idea to just rely on alphabetic sorting of the mark anchor names. We need to verify what implementations do when more than a mark-to-base lookup applies for the same base+mark glyph combination. Does the first or last one wins? I'm not actually sure. Could you try with a test font using hb-shape?

We may want to make sure that, in these conflicting situations, the Glyphs.app predefined order that gives preference to _top and _bottom marks is still enforced, even as we add support to multiple mark anchors per mark glyph.

Reply:
Ok, I see what you mean and I'll investigate. I'll add a test font, as a test case and also run it through hb-shape to see what happens.

Reply:
GPOS lookups are not cumulative, both lookups will be applied meaning the last wins, better verify it with a font, though. However, I’m not sure anyone can depend on a specific lookup ordering here since UFO has no way to explicitly set it, so alphabetic ordering is at least predictable.

Reply:
Thanks Khaled. With the "last one wins" behaviour confirmed, I think I should add back the _top and _bottom ordering to make sure these two are last, so they "win" in case of conflict. (I'm thinking this could get tricky depending on how the marks are sorted between the various lookups...) As a fallback ordering, I'll keep the alphabetical sorting anyway.

Reply:
I’m not sure forcing top and bottom to be last is good idea, this means the designer has no way to override them.

Reply:
Okay, I don't actually have an opinion on this, I defer to @anthrotype to know how important it is to follow Glyphs.app conventions here. Also it's only in case there is ambiguity so we could also just say that it's the designer's problem to resolve the ambiguity. Before this PR, the sorting was much more important because even when everything was non-ambiguous, all except one of the anchor classes for the mark would be discarded. It's not the case anymore, so maybe it doesn't matter anymore to sort _top and _bottom?

Reply:
The use cases where the mark and base/liga/mark glyphs share more than one anchor would be things like contextual positioning, where in some context (or features) one positioning lookup would be used and in other contexts another positioning lookup would be used. This is out of the scope of the markFeatureWriter.

Maybe a warning when more than one anchor is shared would be enough.

Reply:
Thanks Denis, I agree that a warning would probably be enough

Reply:
I added a warning and a matching test.